draft-ietf-tls-esni.md

```diff
@@ -782,10 +777,30 @@ including servers which do not implement this specification.
 
 ### HelloRetryRequest
 
-It is an error for the client to offer ECH before the HelloRetryRequest but not
-after. Likewise, it is an error for the client to offer ECH after the
-HelloRetryRequest but not before. If either of these conditions occurs, then the
-client-facing server MUST abort the handshake with an "illegal_parameter" alert.
+In case a HelloRetryRequest (HRR) is sent, the client-facing server enforces the
+following invariant. Let CH1' denote the first ClientHello processed by the
+server and let CH2' denote the second: either CH1' == CH2' == ClientHelloOuter
+or CH1' == CH2' == ClientHelloInner. In particular, it is an error to process
+ClientHelloInner followed by ClientHelloOuter or vice versa, as this could
+trigger undefined protocol behavior.
+
+Let CH1 and CH2 denote, respectively, the first and second ClientHello
+transmitted on the wire by the client. (Note that these are not necessarily the
+same as the messages processed by the server.) To enforce this invariant, the
+client-facing server behaves as follows.
+
+1. If CH1 contains the "encrypted_client_hello" extension but CH2 does not, or
+   if CH2 contains the "encrypted_client_hello" extension but CH1 does not, then
+   the server MUST abort the handshake with an "illegal_parameter" alert.
+1. Suppose the "encrypted_client_hello" extension is sent in both CH1 and CH2,
+   each extension having a well-formed ClientECH structure as its payload. If
```

Comment:
Nit: I don't think we need to add well-formed ClientECH structure to the precondition. The client-facing server is should already fail the connection with a decode_error if the extension doesn't parse.